### PR TITLE
ref(js): Rewrite ThemeAndStyleProvider as a functional component

### PR DESCRIPTION
**Pull request by evanpurkhiser**
Theme does not need to be kept in state, since we can just rely on the provided config object being updated from the withConfig

### CODE REVIEW

File: static/app/bootstrap/renderMain.tsx [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import {ROOT_ELEMENT} from 'app/constants';
-import Main from 'app/main';
+import Main from 'app/views/main';
```
Comment (evanpurkhiser) [inline static/app/bootstrap/renderMain.tsx:2]: main it doesn't make sense for it to live in views 🤔

File: static/app/components/themeAndStyleProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import {useEffect} from 'react';
+import ReactDOM from 'react-dom';
+import {cache} from '@emotion/css'; // eslint-disable-line emotion/no-vanilla
+import {CacheProvider, ThemeProvider} from '@emotion/react'; // This is needed to set "speedy" = false (for percy)
```
Comment (matejminar) [inline static/app/components/themeAndStyleProvider.tsx:4]: Is this comment still up to date now that we don't use percy?

**Review by matejminar — APPROVED** [review]

File: static/app/components/themeAndStyleProvider.tsx (hunk 2 above)
Comment (evanpurkhiser) [reply static/app/components/themeAndStyleProvider.tsx:4]: good question cc @billyvg

Comment (billyvg) [reply static/app/components/themeAndStyleProvider.tsx:4]: Hmm, I don't remember -- it might be needed for jest snapshots.

**Pull request merged**